### PR TITLE
New panorama pipeline for fisheye images

### PR DESCRIPTION
**Pull request by fabiencastan**
## Description

Add support for HDRI creation from fisheye images.

## Features list

- [X] New Intrinsic: EquidistantRadial
- [X] New dedicated BA with symbolic derivatives

## Implementation remarks

- [X] Change "Intrinsics" inheritance tree and use distortion objects

### CODE REVIEW

File: src/aliceVision/sfm/BundleAdjustmentPanoramaCeres.cpp [hunk 1]
```diff
@@ -0,0 +1,882 @@
...
+    std::size_t nbCamDistUpperOne = 0;
+
+    for(const auto & camdistIt : nbCamerasPerDistance)
+    {
+      if(camdistIt.first < 0)
+        nbCamNotConnected += camdistIt.second;
+      else if(camdistIt.first == 1)
+        nbCamDistEqZero += camdistIt.second;
+      else if(camdistIt.first == 1)
+        nbCamDistEqOne += camdistIt.second;
```
Comment (simogasp) [inline src/aliceVision/sfm/BundleAdjustmentPanoramaCeres.cpp]: As pointed out by codacy, these two conditions are the same. Which one should be removed or what condition should be changed?

File: src/aliceVision/sfm/BundleAdjustmentPanoramaCeres.cpp [hunk 2]
```diff
@@ -0,0 +1,882 @@
...
+    const auto& intrinsicPtr = intrinsicPair.second;
+    const auto usageIt = intrinsicsUsage.find(intrinsicId);
+    if(usageIt == intrinsicsUsage.end()) {
+      // if the intrinsic is never referenced by any view, skip it
+      continue;
+    }
+    const std::size_t usageCount = usageIt->second;
+
+    // do not refine an intrinsic does not used by any reconstructed view
+    if(usageCount <= 0 || getIntrinsicState(intrinsicId) == EParameterState::IGNORED)
```
Comment (simogasp) [inline src/aliceVision/sfm/BundleAdjustmentPanoramaCeres.cpp]: `usageCount` can't be -1. Use `UndefinedIndexT` instead

File: src/aliceVision/camera/Equidistant.hpp [hunk 3]
```diff
@@ -9,175 +9,425 @@
...
+    const double fmm = _scale_x * rscale;
+    const double fov = rsensor / fmm;
+    const double radius = angle_Z / (0.5 * fov);
+
+    /* radius = focal * angle_Z */
+    Vec2 P;
+    P(0) = cos(angle_radial) * radius;
+    P(1) = sin(angle_radial) * radius;
+
+    const Vec2 distorted = this->add_disto(P);
```
Comment (simogasp) [inline src/aliceVision/camera/Equidistant.hpp]: can we remove it as never use it or should that be used in `getDerivativeAddDistoWrtDisto()`?

File: src/aliceVision/camera/Pinhole.hpp [hunk 4]
```diff
@@ -20,202 +20,227 @@
...
+
+    Eigen::Matrix<double, 2, 3> d_P_d_X;
+    d_P_d_X(0, 0) = 1 / X(2);
+    d_P_d_X(0, 1) = 0;
+    d_P_d_X(0, 2) = - X(0) / (X(2) * X(2));
+    d_P_d_X(1, 0) = 0;
+    d_P_d_X(1, 1) = 1 / X(2);
+    d_P_d_X(1, 2) = - X(1) / (X(2) * X(2));
+
+    const Vec2 distorted = this->add_disto(P);
```
Comment (simogasp) [inline src/aliceVision/camera/Pinhole.hpp]: remove it?

File: src/aliceVision/sfm/BundleAdjustmentPanoramaCeres.cpp (hunk 2 above)
Comment (fabiencastan) [reply src/aliceVision/sfm/BundleAdjustmentPanoramaCeres.cpp]: `usageCount` is a counter: so `<= 0` can be replaced by `== 0`. It's not related to `UndefinedIndexT`.

File: src/aliceVision/sfm/BundleAdjustmentPanoramaCeres.cpp (hunk 1 above)
Comment (fabiencastan) [reply src/aliceVision/sfm/BundleAdjustmentPanoramaCeres.cpp]: Yes, obviously it should be:
```
      if(camdistIt.first < 0)
        nbCamNotConnected += camdistIt.second;
      else if(camdistIt.first == 0)  //<<< Should be zero here!
        nbCamDistEqZero += camdistIt.second;
      else if(camdistIt.first == 1)
        nbCamDistEqOne += camdistIt.second;
```

File: src/software/pipeline/main_panoramaEstimation.cpp [hunk 5]
```diff
@@ -42,6 +46,188 @@ inline std::istream& operator>>(std::istream& in, std::pair<int, int>& out)
     return in;
 }
 
+
+
+int main2(int argc, char **argv) {
```
Comment (fabiencastan) [inline src/software/pipeline/main_panoramaEstimation.cpp]: @fabienservant Could we remove this test?

File: src/aliceVision/camera/Equidistant.hpp [hunk 6]
```diff
@@ -9,175 +9,419 @@
...
-  double getFocalLengthPix() const { return _K(0,0); }
+  Vec2 project(const geometry::Pose3& pose, const Vec3& pt, bool applyDistortion = true) const override
+  {    
+    double rsensor = std::min(sensorWidth(), sensorHeight());
+    double rscale = sensorWidth() / std::max(w(), h());
+    double fmm = _scale(0) * rscale;
+    double fov = rsensor / fmm;
 
-  Vec2 getPrincipalPoint() const { return Vec2(_K(0,2), _K(1,2)); }
+    const Vec3 X = pose.rotation() * pt;
```
Comment (simogasp) [inline src/aliceVision/camera/Equidistant.hpp]: Shouldn't the point be brought in the camera reference system? Why only applying the rotation and not the translation?
Am I missing something?

File: src/aliceVision/camera/DistortionRadial.hpp [hunk 7]
```diff
@@ -0,0 +1,596 @@
...
+  static double distoFunctor(const std::vector<double> & params, double r2)
+  {
+    const double k1 = params[0], k2 = params[1], k3 = params[2];
+    return r2 * Square(1.+r2*(k1+r2*(k2+r2*k3)));
+  }
+
+  ~DistortionRadialK3() override = default;
+};
+
+class DistortionRadialK3PT : public Distortion {
```
Comment (simogasp) [inline src/aliceVision/camera/DistortionRadial.hpp]: do u have a reference for this? @fabienservant

File: src/aliceVision/hdr/LaguerreBACalibration.cpp [hunk 8]
```diff
@@ -19,291 +19,284 @@
...
-        {
-            const std::vector<float>& camExp = cameraExposures[i];
-            std::vector<int>& camRelativeExpIndex = closestRelativeExpIndex[i];
-            camRelativeExpIndex.resize(camExp.size(), -1);
-            for(int j = 0; j < camExp.size(); ++j)
+        std::vector<ImageSamples> & group = samples[groupId];
+
+        for (int imageId = 0; imageId < group.size() - 1; imageId++) {
+
+            const ImageSamples & samples = group[imageId];
```
Comment (simogasp) [inline src/aliceVision/hdr/LaguerreBACalibration.cpp]: a variable `samples` is already declared at line 181.
It is better to rename it and even better use the range loops where possible

File: src/aliceVision/camera/DistortionFisheye.hpp [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+    const double theta_dist = theta + k1*theta3 + k2*theta5 + k3*theta7 + k4*theta9;
+    const double cdist = theta_dist / r;
+
+    return p * cdist;
+  }
+
+   Eigen::Matrix2d getDerivativeAddDistoWrtPt(const Vec2 & p) const override
+  {
+      const double eps = 1e-8;
+      const double r = sqrt(p(0) * p(0) + p(1) * p(1));
```
Comment (simogasp) [inline src/aliceVision/camera/DistortionFisheye.hpp:66]: we can use `std::hypot` like in `project()`

File: src/aliceVision/camera/CMakeLists.txt [hunk 10]
```diff
@@ -24,6 +33,7 @@ alicevision_add_interface(aliceVision_camera
 )
 
 # Unit tests
+alicevision_add_test(distortion_test.cpp NAME "camera_distortionRadial"    LINKS aliceVision_camera)
```
Comment (natowi) [inline src/aliceVision/camera/CMakeLists.txt:36]: (cosmetics) align NAME with lines 37-39

Comment (simogasp) [reply src/aliceVision/camera/CMakeLists.txt:36]: yes my bad, i changed the name a couple of times and the formatting lost consistency, fixed

File: src/aliceVision/matchingImageCollection/GeometricFilterMatrix_F_AC.hpp [hunk 11]
```diff
@@ -106,24 +106,38 @@ struct GeometricFilterMatrix_F_AC : public GeometricFilterMatrix
...
       }
       break;
       case robustEstimation::ERobustEstimator::LORANSAC:
       {
         if(m_estimateDistortion)
+        {
           throw std::invalid_argument("["+std::string(__func__)+"] Using fundamental matrix and f10 solver with LO_RANSAC is not yet implemented");
+        }
+        if (cam_I_equidistant && cam_J_equidistant)
+        {
```
Comment (fabiencastan) [inline src/aliceVision/matchingImageCollection/GeometricFilterMatrix_F_AC.hpp:138]: TODO: version for loransac and equidistant

**Pull request merged**